File: library/src/test/java/io/fotoapparat/parameter/selector/LensPositionSelectorsTest.java
package io.fotoapparat.parameter.selector;

import org.junit.Test;

import java.util.List;

import io.fotoapparat.parameter.LensPosition;

import static java.util.Arrays.asList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class LensPositionSelectorsTest {

    @Test
    public void front_Available() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.BACK,
                LensPosition.FRONT,
                LensPosition.EXTERNAL
        );

        // When
        LensPosition result = LensPositionSelectors
                .front()
                .select(availablePositions);

        // Then
        assertEquals(
                LensPosition.FRONT,
                result
        );
    }

    @Test
    public void front_NotAvailable() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.BACK,
                LensPosition.EXTERNAL
        );

        // When
        LensPosition result = LensPositionSelectors
                .front()
                .select(availablePositions);

        // Then
        assertNull(result);
    }

    @Test
    public void back_Available() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.BACK,
                LensPosition.FRONT,
                LensPosition.EXTERNAL
        );

        // When
        LensPosition result = LensPositionSelectors
                .back()
                .select(availablePositions);

        // Then
        assertEquals(
                LensPosition.BACK,
                result
        );
    }

    @Test
    public void back_NotAvailable() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.FRONT,
                LensPosition.EXTERNAL
        );

        // When
        LensPosition result = LensPositionSelectors
                .back()
                .select(availablePositions);

        // Then
        assertNull(result);
    }

    @Test
    public void external_Available() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.BACK,
                LensPosition.FRONT,
                LensPosition.EXTERNAL
        );

        // When
        LensPosition result = LensPositionSelectors
                .external()
                .select(availablePositions);

        // Then
        assertEquals(
                LensPosition.EXTERNAL,
                result
        );
    }

    @Test
    public void external_NotAvailable() throws Exception {
        // Given
        List<LensPosition> availablePositions = asList(
                LensPosition.FRONT,
                LensPosition.BACK
        );

        // When
        LensPosition result = LensPositionSelectors
                .external()
                .select(availablePositions);

        // Then
        assertNull(result);
    }

}

File: fotoapparat/src/test/java/io/fotoapparat/selector/LensPositionSelectorsTest.kt
package io.fotoapparat.selector

import io.fotoapparat.characteristic.LensPosition
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class LensPositionSelectorsTest {

    @Test
    fun `Select front camera which is available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Back,
                LensPosition.Front,
                LensPosition.External
        )

        // When
        val result = front()(availablePositions)

        // Then
        assertEquals(
                LensPosition.Front,
                result
        )
    }

    @Test
    fun `Select front camera which is not available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Back,
                LensPosition.External
        )

        // When
        val result = front()(availablePositions)

        // Then
        assertNull(result)
    }

    @Test
    fun `Select back camera which is available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Back,
                LensPosition.Front,
                LensPosition.External
        )

        // When
        val result = back()(availablePositions)

        // Then
        assertEquals(
                LensPosition.Back,
                result
        )
    }

    @Test
    fun `Select back camera which is not available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Front,
                LensPosition.External
        )

        // When
        val result = back()(availablePositions)

        // Then
        assertNull(result)
    }

    @Test
    fun `Select external camera which is available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Back,
                LensPosition.Front,
                LensPosition.External
        )

        // When
        val result = external()(availablePositions)

        // Then
        assertEquals(
                LensPosition.External,
                result
        )
    }

    @Test
    fun `Select external camera which is not available`() {
        // Given
        val availablePositions = listOf(
                LensPosition.Front,
                LensPosition.Back
        )

        // When
        val result = external()(availablePositions)

        // Then
        assertNull(result)
    }

}